crates/biome_js_analyze/src/lint/nursery/no_adjacent_overload_signatures.rs

```diff
@@ -0,0 +1,332 @@
+use biome_analyze::{context::RuleContext, declare_rule, Ast, Rule, RuleDiagnostic, RuleSource};
+use biome_console::markup;
+use biome_js_syntax::{
+    AnyJsClassMember, AnyJsModuleItem, AnyJsStatement, JsExport, JsIdentifierBinding,
+    JsModuleItemList, JsSyntaxKind, SyntaxNodeText, TsMethodSignatureTypeMember,
+};
+use biome_rowan::{AstNode, TextRange};
+
+declare_rule! {
+    /// Disallow non adjacent overload signatures.
+    ///
+    /// Overload signatures must be adjacent.
+    /// If a key is defined multiple times, only the last definition takes effect. Previous definitions are ignored.
+    /// This rule is useful for preventing accidental overloads that are not adjacent.
+    /// It is recommended to keep the overload signatures adjacent to make the code easier to read and maintain.
+    ///
+    /// Source: https://typescript-eslint.io/rules/adjacent-overload-signatures/
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    ///
+    /// type Foo = {
+    ///   foo_type(s: string): void;
+    ///   foo_type(n: number): void;
+    ///   bar_type(): void;
+    ///   foo_type(sn: string | number): void;
+    /// };
+    ///
+    /// interface Foo {
+    ///   foo_interface(s: string): void;
+    ///   foo_interface(n: number): void;
+    ///   bar_interface(): void;
+    ///   foo_interface(sn: string | number): void;
+    /// }
+    ///
+    /// class A {
+    ///   fooA(s: string): void;
+    ///   fooA(n: number): void;
+    ///   barA(): void {};
+    ///   fooA(sn: string | number): void {};
+    /// }
+    /// ```
```

Comment:
A reminder from the [contribution guide](https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md#document-the-rule):

> When adding invalid snippets in the ### Invalid section, you must use the expect_diagnostic code block property. We use this property to generate a diagnostic and attach it to the snippet. A snippet must emit only ONE diagnostic.

In you case, each `type`/`interface` should be inside a single code block with `expect_diagnostic`